### PR TITLE
Introduce `array_except` function

### LINKED ISSUE #6979: Implement `array_except` function
### Is your feature request related to a problem or challenge?

## Summary

| Characteristic        | Description                                                                                                                                                                                                                                         |
| --------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Function name:        | `array_except`                                                                                                                                                                                                                                      |
| Aliases:              | `list_except`                                                                                                                                                                                                                                       |
| Original function?:   | No                                                                                                                                                                                                                                                  |
| Function Description: | <b>Azure DataBricks:</b> Returns an array of the elements in array1 but not in array2. </br> <b> Spark SQL: </b> Returns all elements from col1 array but not in col2 array.                                                                        |
| Sources:              | [Concept](https://github.com/apache/arrow-datafusion/discussions/6855) [Azure](https://learn.microsoft.com/en-us/azure/databricks/sql/language-manual/functions/array_except) [Spark](https://sparkbyexamples.com/spark/spark-sql-array-functions/) |

## Examples:

```
select array_except([1, 2, 3, 4], [2, 3, 1, 5]);
----
[4, 5]
```

```
select array_except([1, 2, 2, 3, 4], [1, 3, 1, 5]);
----
[2, 4]
```


### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #6979

### CODE REVIEW

File: datafusion/proto/src/logical_plan/from_proto.rs [hunk 1]
```diff
@@ -1342,6 +1347,10 @@ pub fn parse_expr(
                     parse_expr(&args[0], registry)?,
                     parse_expr(&args[1], registry)?,
                 )),
+                ScalarFunction::ArrayIntersect => Ok(array_intersect(
```
Comment (jayzhan211) [inline datafusion/proto/src/logical_plan/from_proto.rs:1350]: I dont why without this can pass CI.

File: datafusion/sqllogictest/test_files/array.slt [hunk 2]
```diff
@@ -2454,6 +2454,113 @@ select list_has_all(make_array(1,2,3), make_array(4,5,6)),
 ----
 false true false true
 
+## array_except
+
+statement ok
```
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/array.slt:2459]: Unlike other tests, I try a new style of test, hope this is easier for review.

File: datafusion/proto/src/logical_plan/from_proto.rs (hunk 1 above)
Comment (Veeupup) [reply datafusion/proto/src/logical_plan/from_proto.rs:1350]: Do we need to enhance the check during CI? maybe this is a little bug?

File: datafusion/sqllogictest/test_files/array.slt (hunk 2 above)
Comment (comphead) [reply datafusion/sqllogictest/test_files/array.slt:2459]: this is choice of style, imho `no-table` tests are easier to read.
```
query ?
select array_expect([], []), array_except(....)
----
[], ...
```

**Review by comphead — COMMENTED** [review]
Thanks @jayzhan211 please also add tests with empty arrays

And since we have introduced the new built in function we need to get it documented in
`scalar_functions.md` and expressions.md`

File: datafusion/sqllogictest/test_files/array.slt (hunk 2 above)
Comment (jayzhan211) [reply datafusion/sqllogictest/test_files/array.slt:2459]: Well, we still need to test multi-rows cases. The new style I mean is not create table in the beginning of this large file and drop table at the end of the file, but when we are done we drop it.

Comment (comphead) [reply datafusion/sqllogictest/test_files/array.slt:2459]: the multirow still possible to do without creating a table 

```
select ... from (select 1 a, 'asdf' b union all select 2 b, 'zxcv' b)
```

thats the matter of style of course.

Comment (jayzhan211) [reply datafusion/sqllogictest/test_files/array.slt:2667]: We cant differentiate empty array and null currently. So we can fix them in another PR

File: docs/source/user-guide/sql/scalar_functions.md [hunk 3]
```diff
@@ -2249,6 +2249,44 @@ array_union(array1, array2)
...
+  Can be a constant, column, or function, and any combination of array operators.
+- **array2**: Array expression.
+  Can be a constant, column, or function, and any combination of array operators.
+
+#### Example
+
+```
+❯ select array_except([1, 2, 3, 4], [5, 6, 3, 4]);
++----------------------------------------------------+
+| array_union([1, 2, 3, 4], [5, 6, 3, 4]);           |
```
Comment (comphead) [inline docs/source/user-guide/sql/scalar_functions.md:2272]: array_except?

File: docs/source/user-guide/sql/scalar_functions.md [hunk 4]
```diff
@@ -2249,6 +2249,44 @@ array_union(array1, array2)
...
++----------------------------------------------------+
+| [1, 2]                                 |
++----------------------------------------------------+
+❯ select array_union([1, 2, 3, 4], [3, 4, 5, 6]);
```
Comment (comphead) [inline docs/source/user-guide/sql/scalar_functions.md:2276]: array_except?

File: datafusion/sqllogictest/test_files/array.slt (hunk 2 above)
Comment (Veeupup) [reply datafusion/sqllogictest/test_files/array.slt:2667]: I also prefer the new style (create/query/drop table statements are organized more tightly). This will make tests more clear.

**Review by Veeupup — APPROVED** [review]

**Review by comphead — APPROVED** [review]
lgtm thanks @jayzhan211 
Please create a followup ticket to differentiate empty array and null, that sounds important

**Comment by alamb** [comment]
There appears to be a non trivial number of conflicts in this PR now

**Comment by alamb** [comment]
I took the liberty of merging up from main to resolve a merge conflict
